File: src/packages/sdk/src/runtime/run-report.ts
/**
 * RUNTIME-104: Deterministic Report Serialization (RunReport v1)
 *
 * Stable report format for downstream processing (explain, diff, history).
 * Mirrors CheckReport conventions for consistency.
 */

import type { EnhancedEvalResult, ExecutionErrorEnvelope } from "./types";

/**
 * RunReport schema version - increment when breaking changes occur
 */
export const RUN_REPORT_SCHEMA_VERSION = "1";

/**
 * Main run report structure
 * Mirrors CheckReport conventions for consistency
 */
export interface RunReport {
	/** Schema version for compatibility */
	schemaVersion: string;
	/** Unique run identifier */
	runId: string;
	/** Run start timestamp */
	startedAt: string;
	/** Run completion timestamp */
	finishedAt: string;
	/** Runtime information */
	runtime: {
		/** Runtime ID */
		id: string;
		/** Project namespace */
		namespace: string;
		/** Project root path */
		projectRoot: string;
	};
	/** Execution results (sorted by testId for determinism) */
	results: RunResult[];
	/** Failures and errors (sorted by testId for determinism) */
	failures: RunFailure[];
	/** Execution summary */
	summary: RunSummary;
	/** Execution configuration */
	config: RunConfig;

	/** Serialize to JSON string */
	toJSON(): string;
}

/**
 * Individual test result
 */
export interface RunResult {
	/** Test specification ID */
	testId: string;
	/** Test specification name */
	testName: string;
	/** File path where test is defined */
	filePath: string;
	/** AST position in file */
	position: { line: number; column: number };
	/** Test input */
	input: string;
	/** Pass/fail determination */
	pass: boolean;
	/** Numeric score (0-100) */
	score: number;
	/** Execution duration in milliseconds */
	durationMs: number;
	/** Test metadata */
	metadata?: Record<string, unknown>;
	/** Test tags */
	tags?: string[];
	/** Assertion results if available */
	assertions?: Array<{
		name: string;
		passed: boolean;
		message?: string;
	}>;
}

/**
 * Failure or error information
 */
export interface RunFailure {
	/** Test specification ID */
	testId: string;
	/** Test specification name */
	testName: string;
	/** File path where test is defined */
	filePath: string;
	/** AST position in file */
	position: { line: number; column: number };
	/** Failure classification */
	classification: "failed" | "error" | "timeout";
	/** Error envelope for errors/timeouts */
	errorEnvelope?: ExecutionErrorEnvelope;
	/** Human-readable error message */
	message: string;
	/** Failure timestamp */
	timestamp: string;
}

/**
 * Execution summary statistics
 */
export interface RunSummary {
	/** Total number of tests */
	total: number;
	/** Number of passed tests */
	passed: number;
	/** Number of failed tests */
	failed: number;
	/** Number of errors */
	errors: number;
	/** Number of timeouts */
	timeouts: number;
	/** Overall pass rate (0-100) */
	passRate: number;
	/** Average score (0-100) */
	averageScore: number;
	/** Total execution duration */
	totalDurationMs: number;
	/** Execution success (no errors/timeouts) */
	success: boolean;
}

/**
 * Execution configuration
 */
export interface RunConfig {
	/** Executor type */
	executorType: string;
	/** Maximum parallel workers */
	maxParallel?: number;
	/** Default timeout in milliseconds */
	defaultTimeout: number;
	/** Environment information */
	environment: {
		nodeVersion: string;
		platform: string;
		arch: string;
	};
}

/**
 * RunReport builder for creating deterministic reports
 */
export class RunReportBuilder {
	private report: Partial<RunReport> = {
		schemaVersion: RUN_REPORT_SCHEMA_VERSION,
		results: [],
		failures: [],
		summary: {
			total: 0,
			passed: 0,
			failed: 0,
			errors: 0,
			timeouts: 0,
			passRate: 0,
			averageScore: 0,
			totalDurationMs: 0,
			success: true,
		},
	};

	/**
	 * Initialize report with basic metadata
	 */
	constructor(
		private runId: string,
		private runtimeInfo: {
			id: string;
			namespace: string;
			projectRoot: string;
		},
	) {
		this.report.startedAt = new Date().toISOString();
		this.report.runId = runId;
		this.report.runtime = runtimeInfo;
	}

	/**
	 * Add a test result to the report
	 */
	addResult(
		testId: string,
		testName: string,
		filePath: string,
		position: { line: number; column: number },
		input: string,
		result: EnhancedEvalResult,
	): void {
		const runResult: RunResult = {
			testId,
			testName,
			filePath,
			position,
			input,
			pass: result.pass,
			score: result.score,
			durationMs: result.durationMs || 0,
			metadata: result.metadata,
			tags: [], // TODO: Extract from spec
			assertions: result.assertions?.map((assertion, index) => ({
				name: assertion.name || `assertion-${index}`,
				passed: assertion.passed,
				message: assertion.message,
			})),
		};

		this.report.results!.push(runResult);

		// Update summary
		this.updateSummary(result);

		// Add to failures if needed
		if (
			!result.pass ||
			result.classification === "error" ||
			result.classification === "timeout"
		) {
			this.addFailure(testId, testName, filePath, position, result);
		}
	}

	/**
	 * Update summary statistics
	 */
	private updateSummary(result: EnhancedEvalResult): void {
		const summary = this.report.summary!;

		summary.total++;
		summary.totalDurationMs += result.durationMs || 0;

		if (result.pass) {
			summary.passed++;
		} else if (result.classification === "error") {
			summary.errors++;
			summary.success = false;
		} else if (result.classification === "timeout") {
			summary.timeouts++;
			summary.success = false;
		} else {
			summary.failed++;
		}

		// Calculate rates and averages
		summary.passRate =
			summary.total > 0 ? (summary.passed / summary.total) * 100 : 0;

		// Average score calculation (excluding errors/timeouts)
		const scoredResults = this.report.results!.filter((r) => r.score > 0);
		summary.averageScore =
			scoredResults.length > 0
				? scoredResults.reduce((sum, r) => sum + r.score, 0) /
					scoredResults.length
				: 0;
	}

	/**
	 * Add a failure to the report
	 */
	private addFailure(
		testId: string,
		testName: string,
		filePath: string,
		position: { line: number; column: number },
		result: EnhancedEvalResult,
	): void {
		const failure: RunFailure = {
			testId,
			testName,
			filePath,
			position,
			classification:
				result.classification === "error"
					? "error"
					: result.classification === "timeout"
						? "timeout"
						: "failed",
			errorEnvelope: result.errorEnvelope,
			message: result.error || "Test failed",
			timestamp: new Date().toISOString(),
		};

		this.report.failures!.push(failure);
	}

	/**
	 * Set execution configuration
	 */
	setConfig(config: Partial<RunConfig>): void {
		this.report.config = {
			executorType: "local",
			defaultTimeout: 30000,
			environment: {
				nodeVersion: process.version,
				platform: process.platform,
				arch: process.arch,
			},
			...config,
		};
	}

	/**
	 * Finalize and return the complete report
	 */
	build(): RunReport {
		// Sort results and failures by testId for determinism
		this.report.results!.sort((a, b) => a.testId.localeCompare(b.testId));
		this.report.failures!.sort((a, b) => a.testId.localeCompare(b.testId));

		// Set completion timestamp
		this.report.finishedAt = new Date().toISOString();

		const finalReport = this.report as RunReport;

		// Add toJSON method
		finalReport.toJSON = () => JSON.stringify(finalReport, null, 2);

		return finalReport;
	}

	/**
	 * Serialize report to JSON string
	 * Ensures deterministic output
	 */
	toJSON(): string {
		const report = this.build();

		return JSON.stringify(report, null, 2);
	}

	/**
	 * Write report to file
	 */
	async writeToFile(filePath: string): Promise<void> {
		const fs = await import("node:fs/promises");
		await fs.writeFile(filePath, this.toJSON(), "utf-8");
	}
}

/**
 * Create a new RunReport builder
 */
export function createRunReport(
	runId: string,
	runtimeInfo: {
		id: string;
		namespace: string;
		projectRoot: string;
	},
): RunReportBuilder {
	return new RunReportBuilder(runId, runtimeInfo);
}

/**
 * Parse a RunReport from JSON string
 */
export function parseRunReport(json: string): RunReport {
	const report = JSON.parse(json) as RunReport;

	// Validate schema version
	if (report.schemaVersion !== RUN_REPORT_SCHEMA_VERSION) {
		throw new Error(
			`Unsupported RunReport schema version: ${report.schemaVersion}`,
		);
	}

	return report;
}

File: src/packages/sdk-python/src/evalgate_sdk/runtime/run_report.py
"""Deterministic RunReport serialization (T4).

Port of the TypeScript SDK's ``run-report.ts``.
Provides a stable report format for downstream processing (explain, diff, history).
"""

from __future__ import annotations

import json
import platform
import sys
from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone
from typing import Any

RUN_REPORT_SCHEMA_VERSION = "1"


@dataclass
class RunResult:
    """Individual test result."""

    test_id: str
    test_name: str
    file_path: str
    position: dict[str, int]
    input: str
    passed: bool
    score: float
    duration_ms: float
    metadata: dict[str, Any] | None = None
    tags: list[str] = field(default_factory=list)
    assertions: list[dict[str, Any]] = field(default_factory=list)


@dataclass
class RunFailure:
    """Failure or error information."""

    test_id: str
    test_name: str
    file_path: str
    position: dict[str, int]
    classification: str  # "failed" | "error" | "timeout"
    message: str
    timestamp: str = ""
    error_envelope: dict[str, Any] | None = None


@dataclass
class RunSummary:
    """Execution summary statistics."""

    total: int = 0
    passed: int = 0
    failed: int = 0
    errors: int = 0
    timeouts: int = 0
    pass_rate: float = 0.0
    average_score: float = 0.0
    total_duration_ms: float = 0.0
    success: bool = True


@dataclass
class RunConfig:
    """Execution configuration."""

    executor_type: str = "local"
    max_parallel: int | None = None
    default_timeout: int = 30_000
    environment: dict[str, str] = field(default_factory=dict)


@dataclass
class RunReport:
    """Main run report structure."""

    schema_version: str = RUN_REPORT_SCHEMA_VERSION
    run_id: str = ""
    started_at: str = ""
    finished_at: str = ""
    runtime: dict[str, str] = field(default_factory=dict)
    results: list[RunResult] = field(default_factory=list)
    failures: list[RunFailure] = field(default_factory=list)
    summary: RunSummary = field(default_factory=RunSummary)
    config: RunConfig = field(default_factory=RunConfig)

    def to_json(self) -> str:
        """Serialize to deterministic JSON string."""
        return json.dumps(asdict(self), indent=2, default=str)


class RunReportBuilder:
    """Builder for creating deterministic RunReport instances."""

    def __init__(
        self,
        run_id: str,
        runtime_info: dict[str, str],
    ) -> None:
        self._report = RunReport(
            run_id=run_id,
            started_at=datetime.now(timezone.utc).isoformat(),
            runtime=runtime_info,
            config=RunConfig(
                environment={
                    "python_version": sys.version.split()[0],
                    "platform": platform.system().lower(),
                    "arch": platform.machine(),
                },
            ),
        )
        self._scores: list[float] = []

    def add_result(
        self,
        test_id: str,
        test_name: str,
        file_path: str,
        position: dict[str, int],
        input: str,
        *,
        passed: bool,
        score: float,
        duration_ms: float = 0.0,
        metadata: dict[str, Any] | None = None,
        tags: list[str] | None = None,
        assertions: list[dict[str, Any]] | None = None,
        classification: str = "passed",
        error: str | None = None,
        error_envelope: dict[str, Any] | None = None,
    ) -> None:
        """Add a test result to the report."""
        result = RunResult(
            test_id=test_id,
            test_name=test_name,
            file_path=file_path,
            position=position,
            input=input,
            passed=passed,
            score=score,
            duration_ms=duration_ms,
            metadata=metadata,
            tags=tags or [],
            assertions=assertions or [],
        )
        self._report.results.append(result)

        # Update summary
        s = self._report.summary
        s.total += 1
        s.total_duration_ms += duration_ms

        if passed:
            s.passed += 1
        elif classification == "error":
            s.errors += 1
            s.success = False
        elif classification == "timeout":
            s.timeouts += 1
            s.success = False
        else:
            s.failed += 1
            s.success = False

        s.pass_rate = (s.passed / s.total * 100) if s.total > 0 else 0.0
        if score > 0:
            self._scores.append(score)
        s.average_score = (sum(self._scores) / len(self._scores)) if self._scores else 0.0

        # Add to failures if needed
        if not passed or classification in ("error", "timeout"):
            failure = RunFailure(
                test_id=test_id,
                test_name=test_name,
                file_path=file_path,
                position=position,
                classification=classification if classification in ("error", "timeout") else "failed",
                message=error or "Test failed",
                timestamp=datetime.now(timezone.utc).isoformat(),
                error_envelope=error_envelope,
            )
            self._report.failures.append(failure)

    def set_config(self, **kwargs: Any) -> None:
        """Update execution configuration fields."""
        for key, value in kwargs.items():
            if hasattr(self._report.config, key):
                setattr(self._report.config, key, value)

    def build(self) -> RunReport:
        """Finalize and return the complete report."""
        # Sort for determinism
        self._report.results.sort(key=lambda r: r.test_id)
        self._report.failures.sort(key=lambda f: f.test_id)
        self._report.finished_at = datetime.now(timezone.utc).isoformat()
        return self._report

    def to_json(self) -> str:
        """Build and serialize to JSON."""
        return self.build().to_json()

    async def write_to_file(self, file_path: str) -> None:
        """Write report to file."""
        from pathlib import Path

        Path(file_path).write_text(self.to_json(), encoding="utf-8")


def create_run_report(
    run_id: str,
    runtime_info: dict[str, str],
) -> RunReportBuilder:
    """Create a new RunReport builder."""
    return RunReportBuilder(run_id, runtime_info)


def _filter_dataclass_fields(cls: type, d: dict[str, Any]) -> dict[str, Any]:
    """Keep only keys that match dataclass field names — prevents TypeError on extras."""
    import dataclasses

    valid = {f.name for f in dataclasses.fields(cls)}
    return {k: v for k, v in d.items() if k in valid}


def parse_run_report(json_str: str) -> RunReport:
    """Parse a RunReport from a JSON string."""
    data = json.loads(json_str)
    version = data.get("schema_version", "")
    if version != RUN_REPORT_SCHEMA_VERSION:
        raise ValueError(f"Unsupported RunReport schema version: {version}")

    summary = RunSummary(**_filter_dataclass_fields(RunSummary, data.get("summary", {})))
    config = RunConfig(**_filter_dataclass_fields(RunConfig, data.get("config", {})))
    results = [RunResult(**_filter_dataclass_fields(RunResult, r)) for r in data.get("results", [])]
    failures = [RunFailure(**_filter_dataclass_fields(RunFailure, f)) for f in data.get("failures", [])]

    return RunReport(
        schema_version=data["schema_version"],
        run_id=data["run_id"],
        started_at=data.get("started_at", ""),
        finished_at=data.get("finished_at", ""),
        runtime=data.get("runtime", {}),
        results=results,
        failures=failures,
        summary=summary,
        config=config,
    )
